pkg/plugin.go

```diff
@@ -79,25 +79,32 @@ func newDatasource() datasource.ServeOpts {
 }
 
 // QueryData handles multiple queries and returns multiple responses.
-// req contains the queries []DataQuery (where each query contains RefID as a unique identifer).
+// req contains the queries []DataQuery (where each query contains RefID as a unique identifier).
 // The QueryDataResponse contains a map of RefID to the response for each query, and each response
 // contains Frames ([]*Frame).
 func (td *SampleDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
-	// create response struct
 	response := backend.NewQueryDataResponse()
+	results := make(chan func(backend.Responses), len(req.Queries))
 
-	// loop over queries and execute them individually.
+	// Execute the queries in parallel and collect the responses.
 	for _, q := range req.Queries {
-		res := td.query(ctx, q)
-		// save the response in a hashmap
-		// based on with RefID as identifier
-		response.Responses[q.RefID] = res
+		go func(dq backend.DataQuery) {
+			res := doQuery(ctx, dq)
+			results <- func(r backend.Responses) {
```

Comment:
something irks me about passing a closure to a channel...but I don't have anything concrete.

Reply:
I also considered mutex'ing it, or using a struct, I'm pretty impartial I just didn't want to mutex it. 

![image](https://user-images.githubusercontent.com/21345430/105888323-d2424600-5fda-11eb-8160-dadf5e47ba8d.png)

Reply:
the channel is fine really, if you never get enough results though it will block right?  and I'd like to see the context handling in the for loop.

Reply:
it should still push the correct amount of functions into the channel, unless there's a panic but that obviously stops everything anyway